rust-runtime/aws-smithy-runtime-api/src/http/error.rs

```diff
@@ -16,48 +16,113 @@ use std::str::Utf8Error;
 /// An error occurred constructing an Http Request.
 ///
 /// This is normally due to configuration issues, internal SDK bugs, or other user error.
-pub struct HttpError(BoxError);
+pub struct HttpError {
+    kind: Kind,
+    source: Option<BoxError>,
+}
+
+#[derive(Debug)]
+enum Kind {
+    InvalidExtensions,
+    InvalidHeaderName,
+    InvalidHeaderValue,
+    InvalidStatusCode,
+    InvalidUri,
+    InvalidUriParts,
+    MissingAuthority,
+    MissingScheme,
+    NotUtf8(Vec<u8>),
```

Comment:
why not also include the decoding error here instead of reconstructing it during display?